fabric-api-base/src/main/java/net/fabricmc/fabric/impl/base/event/SortableNode.java

```diff
@@ -14,10 +14,15 @@
  * limitations under the License.
  */
 
-package net.fabricmc.fabric.impl.resource.loader;
+package net.fabricmc.fabric.impl.base.event;
 
-import net.minecraft.resource.ResourceType;
+import java.util.ArrayList;
+import java.util.List;
 
-public interface FabricLifecycledResourceManager {
-	ResourceType fabric_getResourceType();
+public abstract class SortableNode<N extends SortableNode<N>> implements Comparable<N> {
+	public final List<N> subsequentNodes = new ArrayList<>();
+	public final List<N> previousNodes = new ArrayList<>();
+	int visitStatus = 0; // 0: not visited, 1: visiting, 2: visited
+
+	protected abstract void printCycleWarning(N otherNode);
```

Comment:
Actually it would be better to print all the SCCs with size > 1 in the `NodeSorting` method directly. Maybe it can have what it is sorting as a parameter (e.g. `"Found cycle(s) when sorting %s:"` where `%s` could be either `"Event phases"` or `"Resource reloaders"`).